src/ffi/android.rs

```diff
@@ -49,185 +47,75 @@ pub unsafe extern "system" fn Java_org_c4dt_artiwrapper_TorLibApi_initLogger(_:
-                    entry.into(),
-                ],
-            )
-            .context("call HashMap.put()")?
-        {
-            JValue::Object(o) => {
-                if o.into_inner() != std::ptr::null_mut() {
-                    trace!("Entry already existed -- appending");
-                    entry = o;
-                }
-            }
-            _ => (),
-        }
-
-        env.call_method(
-            entry,
-            "add",
-            "(Ljava/lang/Object;)Z",
-            &[env
-                .new_string(value.to_str().context("convert header value to string")?)
-                .context("create JString for value")?
-                .into()],
-        )
-        .context("call List.add()")?;
-    }
-
-    let body = env
-        .byte_array_from_slice(response.body())
-        .context("create byte array")?;
-
-    Ok(env
-        .new_object(
-            env.find_class("org/c4dt/artiwrapper/HttpResponse")
-                .context("find HttpResult class")?,
-            "(ILjava/lang/String;Ljava/util/Map;[B)V",
-            &[status.into(), version.into(), headers.into(), body.into()],
-        )
-        .context("create HttpResult")?)
+    _: JClass,
+    cache_dir_j: JString,
+) -> jlong {
+    throw_on_err(env, 0, || {
+        // TODO avoid UTF-8 conversion?
+        let cache_dir_javastr = env
+            .get_string(cache_dir_j)
+            .context("create rust string for `cache_dir_j`")?;
+        let cache_dir = cache_dir_javastr
+            .deref()
+            .to_str()
+            .context("rust string from java")
+            .map(Path::new)?;
+
+        RuntimeAndClient::new(cache_dir)
+            .context("create runtime and client")
+            .map(Into::into)
+    })
 }
 
-/// Entry point to process an HTTP request via Arti
+/// Send a request with the given Client
 #[no_mangle]
-pub unsafe extern "system" fn Java_org_c4dt_artiwrapper_TorLibApi_torRequest(
+pub unsafe extern "system" fn Java_org_c4dt_artiwrapper_Client_send(
     env: JNIEnv,
     _: JClass,
-    cache_dir_j: JString,
+    java_client: jlong,
     method_j: JString,
     url_j: JString,
     headers_j: JObject,
     body_j: jbyteArray,
 ) -> jobject {
-    panic::catch_unwind(|| {
-        || -> Result<JObject> {
-            let request =
-                make_request(env, method_j, url_j, headers_j, body_j).context("make request")?;
-            trace!("Request: {:?}", request);
-
-            let cache_dir = get_cache_dir(env, cache_dir_j).context("get cache dir")?;
-            let client = Client::new(PathBuf::from(cache_dir));
+    throw_on_err(env, ptr::null_mut(), || {
+        let rt_and_client = RuntimeAndClient::from(java_client);
+        let request = Request::from_java(env, method_j, url_j, headers_j, body_j)
+            .context("request from java")?;
+
+        let response = rt_and_client
+            .runtime()
+            .block_on(async { rt_and_client.client().send(request.0).await })
```

Comment:
Generic question on `block_on`: are you sure there is nothing else running that might be blocked by this? I had troubles with `block_on` while waiting for a webrtc-answer, because the rest of the package didn't continue.

So if there is anything else running with the runtime, we need to make sure it doesn't block while waiting here.

Reply:
> Generic question on `block_on`: are you sure there is nothing else running that might be blocked by this?

I don't quite see how something else might be running as the runtime is single threaded.

> So if there is anything else running with the runtime, we need to make sure it doesn't block while waiting here.

The only reference to the runtime is in this method.